### PR TITLE
Split data into MemBacked & Filebacked as single file

### PR DESCRIPTION
**Pull request by oerc0122**
First implementation of single-filebacked through `memmapfile`. Currently serial.

Further optimisation to be done for operations (dynamically dumping pixels, rather than pre-dumping) and some other operations.

### CODE REVIEW

File: horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m [hunk 1]
```diff
@@ -1,35 +1,66 @@
...
         % Input
         % -----
         % pix_id   The ID of the PixelData instance this object is linked to.
         %          This sets the tmp directory name
-        % has_tmp_files  Logical array saying if tmp exist for the given page
-        obj.pix_id_ = pix_id;
-        obj.tmp_dir_path_ = obj.generate_tmp_dir_path_(obj.pix_id_);
-        if exist('has_tmp_files', 'var')
-            obj.has_tmp_file_ = has_tmp_files;
+        % has_tmp_file  Logical array saying if tmp exist for the given page
```
Comment (cmarooney-stfc) [inline horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m:41]: there is no indication what kind of id a pix_id should be. Filename? PixelData handle? Datum in an sqw? Needs more info

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

File: horace_core/sqw/PixelData/@PixelDataMemory/append.m [hunk 2]
```diff
@@ -0,0 +1,44 @@
+function pix_out = append(obj, pix)
+% Join the pixels in the given PixelData object to the end of this
+% object.
+%
+% The pixels to append must all be in memory and you cannot append pixels if
+% the inputted PixelData object has more pixels than are allowed in a single
+% page.
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply horace_core/sqw/PixelData/@PixelDataMemory/append.m:7]: This is leftover old documentation, this is no longer the case. Removed

File: horace_core/sqw/@sqw/sqw.m [hunk 3]
```diff
@@ -97,6 +105,18 @@
         varargout = multifit (varargin);
 
 
+        % TOBYFIT interface
+        %------------------------------------------------------------------
+        %TODO: Something in this interface looks dodgy. Should it be just
+        %      TOBYFIT interface, or should it go out of here?
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply horace_core/sqw/@sqw/sqw.m]: Not part of my documentation, not sure where it came from.

File: horace_core/algorithms/read_horace.m [hunk 4]
```diff
@@ -25,58 +25,55 @@
...
-%
+
 if get_dnd && force_sqw
     error('HORACE:read_horace:invalid_argument',...
-        'only one option i.e. -get_dnd or -force_sqw can be provided simultaniously');
+        'only one option i.e. -get_dnd or -force_sqw can be provided simultaneously');
+end
+
+if ~any(strcmp(argi, 'file_backed')) % Prioritise explicit, but otherwise respect -file_backed
+    argi = [argi, 'file_backed', file_backed];
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply horace_core/algorithms/read_horace.m:49]: Yes, rephrased. If `file_backed` specified as kwarg (e.g. `'file_backed', false`) it will not override the user, but will attempt to file back in the case of the `-file_backed` flag (which sets the variable `file_backed`).

File: horace_core/algorithms/read_horace.m [hunk 5]
```diff
@@ -25,58 +25,55 @@
...
-            error('HORACE:read_horace:invalid_argument',...
-                'File %s contains dnd information but only sqw file requested',...
-                fullfile(loaders{i}.filepath,loaders{i}.filename));
-        end
+    is_not_sqw = ~cellfun(@(x) x.sqw_type, loaders);
+    if any(is_not_sqw)
+        bad_files = cellfun(@(x) fullfile(x.filepath,x.filename), ...
+                            loaders(is_not_sqw), 'UniformOutput', false);
+        error('HORACE:read_horace:invalid_argument',...
+              'File %s contains dnd information but only sqw file requested.',...
```
[the remark this answers is not in the document]
Comment (oerc0122) [reply horace_core/algorithms/read_horace.m:61]: Fixed.

File: horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m (hunk 1 above)
Comment (oerc0122) [reply horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m:41]: This is a random number generated by the FileBacked entity in the range [1, 99999] (could get collisions and may want to fix this, but this is how it was done originally)
Old (Computes 5 random numbers and raises each to 10^index for some reason):
```
            obj.object_id_ = polyval(randi([0, 9], 1, 5), 10);
```
New (Computes 1 random number in range 1,99999):
```
            obj.object_id_ = randi([1, 99999], 1, 1);
```

File: horace_core/sqw/@sqw/sqw.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/@sqw/sqw.m]: That's my comment done ages ago. Do not know why it has been picked up here. But indeed, first, methods depend on kind of instrument while they should have instrument-kind oblivious form and do their own thing according to overloading. 

Second -- the question it they should be there at all.

File: herbert_core/utilities/files/do_fseek.m [hunk 6]
```diff
@@ -0,0 +1,16 @@
+function status = do_fseek(fid, offset, origin)
+%%DO_FSEEK Call fseek with the given arguments, throw an error if something goes wrong
```
Comment (abuts) [inline herbert_core/utilities/files/do_fseek.m:2]: `fseek` does not thrown in Matlab, so `try/catch` is in mainly useless.  A lot of unnecessary changes to extract this, but even more unnecessary efforts to revert

Comment (oerc0122) [reply herbert_core/utilities/files/do_fseek.m:2]: Afraid it does:
```
>> q = fopen('dumpy', 'w')
>> try
fseek(q, -1000, 'bot')
catch ME
ME
end

ME =

  MException with properties:

    identifier: 'MATLAB:badorigin_mx'
       message: 'Invalid origin.'
         cause: {}
         stack: [0x1 struct]
    Correction: []

>> fseek(7, 1, 'bof')
Error using fseek
Invalid file identifier.  Use fopen to generate a valid file identifier.

```

Comment (abuts) [reply herbert_core/utilities/files/do_fseek.m:2]: OK then. Either my knowledge is from very old Matlab version or I used it in the situations, where input parameters were verified in different place. As separate routine its necessary

**Review by abuts — APPROVED** [review]
looks like a lot of changes, difficult to judge from simple comparison. Passes on my machine. I need these changes, so will investigate more thoroughly when start using it. Should be merged as soon as Matlab licenses issues are resolved (request submitted) and at least one Jenkins built is successful

**Comment by abuts** [comment]
looks like a lot of changes, difficult to judge from simple comparison. I need these changes, so will investigate more thoroughly when start using it. Should be merged as soon as Matlab licenses

**Review by abuts — CHANGES_REQUESTED** [review]
It looks like `memmapfile` concept is incomplete. Tests are failing which are intended for verifying paging sub-files and this have not been re-implemented and the concept is incomplete. 

I am still working on the attempt to separate this on `memmapfile` (incomplete) and split (useful and ready)
so further work is necessary

File: horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m (hunk 1 above)
Comment (cmarooney-stfc) [reply horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m:41]: Probably an excellent explanation but it has not been put in the file. IF this information exists elsewhere, provide a location

Comment (oerc0122) [reply horace_core/sqw/PixelData/@PixelTmpFileHandler/PixelTmpFileHandler.m:41]: It is documented:
```
        % pix_id   The ID of the PixelData instance this object is linked to.
        %          This sets the tmp directory name
```
And on PixelDataBase
```
        object_id_;  % random unique identifier for this object, used for tmp file names
```

**Review by abuts — COMMENTED** [review]
It is certainly incomplete. 
`PixelTmpFileHandler` tries to delete file which never exist

**Review by abuts — APPROVED** [review]
code is broken, and I do not see all changes promised. But as part of the process of splitting, and failures are highlighted -- the step in right direction

File: _test/test_sqw_pixels/test_PixelData.m [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
     properties
 
-        SMALL_PG_SIZE = 1e6;  % 1Mb
+        SMALL_PG_SIZE = 1e5;  % 100,000 pix % This disables filebased access and makes tested files memory based. Not what is intended in tests
```
Comment (cmarooney-stfc) [inline _test/test_sqw_pixels/test_PixelData.m:5]: Is this saying that the test does not work? Guess we should rejig the test then. If it does work then maybe the comment should say something different.

File: _test/test_sqw_pixels/test_pixels_equal.m [hunk 8]
```diff
@@ -21,11 +21,12 @@
             pths = horace_paths;
             obj.test_sqw_file_path = fullfile(pths.test_common, 'sqw_2d_1.sqw');
 
-            % sqw_2d_1.sqw has ~1.8 MB of pixels, a 400 kB page size gives us 5
+            % sqw_2d_1.sqw has ~25,000 pixels, a 5000 pixels gives us 5
```
Comment (cmarooney-stfc) [inline _test/test_sqw_pixels/test_pixels_equal.m:24]:
```suggestion
            % sqw_2d_1.sqw has ~25,000 pixels, at 5000 pixels per page gives us 5
```

**Review by cmarooney-stfc — CHANGES_REQUESTED** [review]

**Comment by abuts** [comment]
As tests are passing and I need to follow up these changes, I am happy to merge them
